kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -30,157 +32,143 @@ import (
-
-	listFunc := func() (map[string]interface{}, error) {
-		log.Debugf("Listing profiles from Calico datastore")
-		filteredServices := make(map[string]interface{})
-
-		// Get all profile objects from Calico datastore.
-		serviceList, err := clientset.CoreV1().Services("").List(ctx, metav1.ListOptions{})
-		if err != nil {
-			return nil, err
-		}
-
-		// Filter out only services of type LoadBalancer
-		for _, svc := range serviceList.Items {
-			if svc.Spec.Type == v1.ServiceTypeLoadBalancer {
-				key := createHandle(&svc)
-				filteredServices[key] = svc
-			}
-		}
-		log.Debugf("Found %d Service LoadBalancer in Calico datastore", len(filteredServices))
-		return filteredServices, nil
-	}
-
-	cacheArgs := rcache.ResourceCacheArgs{
-		ListFunc:    listFunc,
-		ObjectType:  reflect.TypeOf(&v1.Service{}),
-		LogTypeDesc: "Service",
-	}
-	ccache := rcache.NewResourceCache(cacheArgs)
-
-	svcHandler := cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			log.Debugf("Got ADD event for service")
-			svc := obj.(*v1.Service)
-			if isCalicoManagedLoadBalancer(svc, cfg.AssignIPs) {
-				ccache.Set(createHandle(svc), svc)
+func NewLoadBalancerController(clientset *kubernetes.Clientset, calicoClient client.Interface, cfg config.LoadBalancerControllerConfig, serviceInformer cache.SharedIndexInformer, dataFeed *node.DataFeed) controller.Controller {
+	lbc := &loadBalancerController{
+		calicoClient:     calicoClient,
+		cfg:              cfg,
+		clientSet:        clientset,
+		dataFeed:         dataFeed,
+		syncerUpdates:    make(chan interface{}),
+		syncChan:         make(chan interface{}, 1),
+		ipamBlocks:       make(map[string]model.KVPair),
+		ipPools:          make(map[string]api.IPPool),
+		serviceInformer:  serviceInformer,
+		serviceLister:    v1lister.NewServiceLister(serviceInformer.GetIndexer()),
+		servicesToUpdate: make(map[string]serviceObject),
+	}
+
+	lbc.RegisterWith(lbc.dataFeed)
+	lbc.dataFeed.Start()
+
+	_, err := lbc.serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: func(obj interface{}) {
+			if svc, ok := obj.(*v1.Service); ok {
+				if IsCalicoManagedLoadBalancer(svc, cfg.AssignIPs) {
+					lbc.syncerUpdates <- serviceObject{
+						updateType: serviceUpdateTypeDELETE,
+					}
+				}
 			}
 		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
-			log.Debugf("Got UPDATE event for service")
-			log.Debugf("Old object: %v", oldObj)
-			log.Debugf("New object: %v", newObj)
 			svcNew := newObj.(*v1.Service)
 			svcOld := oldObj.(*v1.Service)
-
-			// Service changed type, we need to release the addresses used by the service
+			handle, err := createHandle(svcNew)
+			if err != nil {
+				log.WithError(err).Error("Error creating load balancer handle")
+				return
+			}
 			if svcNew.Spec.Type != v1.ServiceTypeLoadBalancer &&
 				svcOld.Spec.Type == v1.ServiceTypeLoadBalancer {
-				ccache.Delete(createHandle(svcOld))
+				lbc.syncerUpdates <- serviceObject{
```

Comment:
`syncerUpdates` implies that the updates are coming from the syncer, but these are actually coming from the informer